src/cargo/core/dependency.rs

```diff
@@ -109,18 +108,25 @@ This will soon become a hard error, so it's either recommended to
 update to a fixed version or contact the upstream maintainer about
 this warning.
 ",
-                        req,
-                        inside.name(),
-                        inside.version(),
-                        requirement
-                    );
-                    config.shell().warn(&msg)?;
-
-                    Ok(requirement)
-                }
-                e => Err(e.into()),
-            }
-        }
+                req,
+                inside.name(),
+                inside.version(),
+                requirement
+            );
+            config.shell().warn(&msg)?;
+
+            Ok(requirement)
+        },
+        Err(e) => {
+            let err: CargoResult<VersionReq> = Err(e.into());
+            let v: VersionReq = err.chain_err(|| {
+                format!(
+                    "failed to parse the version requirement `{}` for dependency `{}`",
+                    req, name
+                )
+            })?;
+            Ok(v)
```

Comment:
This is just confusing.  The "CargoResult" is always an `Err` so `?` will never result in an actual `VersionReq`.